ext/openssl/ossl_x509cert.c

```diff
@@ -704,6 +704,188 @@ ossl_x509_eq(VALUE self, VALUE other)
+            ossl_clear_error();
+        }
+
+        if (ERR_peek_last_error())
+            ossl_raise(eX509CertError, NULL);
+        else
+            return Qnil;
+    }
+
+    certificates = load_chained_certificates_append(Qnil, certificate);
+
+    while ((certificate = PEM_read_bio_X509(in, NULL, NULL, NULL))) {
+      load_chained_certificates_append(certificates, certificate);
+    }
+
+    /* We tried to read one more certificate but could not read start line: */
+    if (ERR_GET_REASON(ERR_peek_last_error()) == PEM_R_NO_START_LINE) {
+        /* This is not an error, it means we are finished: */
+        ossl_clear_error();
+
+        return certificates;
+    }
+
+    /* Alternatively, if we reached the end of the file and there was no error: */
+    if (BIO_eof(in) && !ERR_peek_last_error()) {
+        return certificates;
+    } else {
+        /* Otherwise, we tried to read a certificate but failed somewhere: */
+        ossl_raise(eX509CertError, NULL);
+    }
+}
+
+static VALUE
+load_chained_certificates_DER(BIO *in) {
+    X509 *certificate = d2i_X509_bio(in, NULL);
+
+    /* If we cannot read one certificate: */
+    if (certificate == NULL) {
+        if (BIO_eof(in)) {
+            /* If we didn't actually read anything: */
+            if (BIO_tell(in) == 0) {
+                ossl_clear_error();
+            }
+        }
+
+        if (ERR_peek_last_error())
+            ossl_raise(eX509CertError, NULL);
+        else
+            return Qnil;
+    }
+
+    return load_chained_certificates_append(Qnil, certificate);
+}
+
+static VALUE
+load_chained_certificates(VALUE _io) {
+    BIO *in = (BIO*)_io;
+    VALUE certificates = Qnil;
+
+    /* Case 1: certificate format is PEM: */
+    certificates = load_chained_certificates_PEM(in);
+
+    if (certificates != Qnil)
+        return certificates;
+
+    /* Case 2: certificate format is DER: */
+    OSSL_BIO_reset(in);
+
+    certificates = load_chained_certificates_DER(in);
+
+    if (certificates != Qnil)
+        return certificates;
+
+    /* Otherwise we couldn't read the output correctly so fail: */
+    ossl_raise(eX509CertError, "Could not detect format of certificate data!");
+}
+
+static VALUE
+load_chained_certificates_ensure(VALUE _io) {
+    BIO *in = (BIO*)_io;
+
+    BIO_free(in);
+
+    return Qnil;
+}
+
+/*
+ * call-seq:
+ *    OpenSSL::X509::Certificate.load_file(path) -> [certs...]
+ *
+ * Read the chained certificates from specified file path. See
+ * +OpenSSL::X509::Certificate.load+ for more details.
+ */
+static VALUE
+ossl_x509_load_file(VALUE klass, VALUE path)
+{
+    BIO *in = BIO_new(BIO_s_file());
+
+    if (in == NULL)
+        ossl_raise(eX509CertError, NULL);
+
+    if (BIO_read_filename(in, StringValueCStr(path)) <= 0) {
+        BIO_free(in);
+        ossl_raise(eX509CertError, NULL);
+    }
+
+    return rb_ensure(load_chained_certificates, (VALUE)in, load_chained_certificates_ensure, (VALUE)in);
```

Comment:
Can `load_file` be implemented around `load`?

```ruby
def load_file(path)
  load(File.read(path))
end
```

All loaded certificates will be on memory anyway after successful loading, so reading all content in the file should not cause any issue.

Reply:
It's less efficient because it will allocate Ruby string. However, it's acceptable to me.